File: 현지/2026_05_28/디펜스 게임.java
import java.util.*;

class Solution {

    public int solution(int n, int k, int[] enemy) {

        int answer = 0;

        PriorityQueue<Integer> q = new PriorityQueue<>();

        for(int i = 0; i < enemy.length; i++) {

            q.offer(enemy[i]);

            // 무적권 넘으면
            if(q.size() > k) {

                int a = q.poll();

                n = n - a;
            }

            // 병사 없으면
            if(n < 0) {

                answer = i;

                return answer;
            }
        }

        answer = enemy.length;

        return answer;
    }
}

File: 현지/2026_05_28/디펜스 게임.py
import heapq

def solution(n, k, enemy):

    h = []
    answer = 0

    for i in range(0, len(enemy)):

        # 일단 넣기
        heapq.heappush(h, enemy[i])

        # k보다 많으면
        if len(h) > k:

            x = heapq.heappop(h)

            n = n - x

        # 병사 없으면 끝
        if n < 0:

            answer = i
            return answer

    answer = len(enemy)

    return answer
